### ISSUE TITLE
No Compiler Warning when Nullability mismatches on a Generic struct being passed via an Interface

### ISSUE THREAD

**Issue by Brondahl** [dotnet/roslyn#79845]
```
 IImmutableList<object> moreNoNulls = new ImmutableArray<object?>();
```

This should give a warning (CS8619)... but doesn't.
Appears to be true for any generic struct implementing a generic interface.
Fails for gen. argument of any reference type; works fine if the generic arg is a struct (which makes sense)

**Steps to Reproduce**:

https://lab.razor.fyi/#jVJNbtQwFBaISuAV4gRWNiRomlIW_E2bqs20VVGnHZEWJBASjudNanDsyM-ZIUU9ASsWbNhyDQ7CLbgCSjKTmUxL1WxsP38_730O-blCyMDoxLDU5_jg20qOQiU0KtBC6odaSuBWaIX-PigwgnfJfxEHaZpbFkvoEsIlQ6RTZfKVUEqpUBbMiHGgB_3iYHbY0LmlJ0GXVBi0JueW9ouo2mycBPRlG95Aa4t-EZbrdUC0zApOx1oMaZ8J5XpVuW6q_NrtxJ-A24AyukkVTBqDur4VuF6Xrq3RngakiRgDnTCjykTC6PnT9RfXi8aN6HS-K1TVQ9sWduPcUjzTuRx6ZK5_KNBuTBMOqNJHuZQ41V-8m4lf6rhDY-AsR6jhARVYh-rPTZo3bbul2sBRy7EBbhvDimXvIz0z7TambUZpXj9-ZX5BLsijXyuhViOR-GF0xkw2YAbhOKt-NlfXK90M6HRb8fy3wp4dMpXkLIE3YFBo5S6d_YGBsYCJN2cMDGRGc0DUJirSWEt0nd7uzum-s4DaA2ZzA-hWpfdNSgomrpMjrBqNslCrVn8GJc7BOB3qWJOD43Xa4D0hYYchDKc5LQM_eMTrEtIaP9RpJiQrZ71JCNtS6smpQjYCt5RemKN8tzL4UCsLX-xM7eqyv6vK4gK9vEjFedXJIYxBupcqfg_iPFngRBlwMRK8J1iiNFrBZ65XZRlG688erzsd-hoybeyc5Ed5lhlAXA60Yjy5AaOd7G4qmvGvC7PE9cGyIbPsWMnCHTGJUEm9us3x3b27v__8_fF9dP_Ox1v_AA

```
using System.Collections.Generic;
using System.Collections.Immutable;

class Program
{
    interface IMyInterface<out T>;

    struct MyStruct<T> : IMyInterface<T>;

    class MyClass<T> : IMyInterface<T>;

    static void Main()
    {
        IMyInterface<object> a = new MyClass<object?>(); // Does give warning CS8619
        IMyInterface<object> b = new MyStruct<object?>(); // Doesn't give warning (but should)

        IList<Program> noNulls = new List<Program?>(); //Does give warning, because List<> is class.
        IImmutableList<Program> moreNoNulls = new ImmutableArray<Program?>(); //No warning; because ImmutableArray is struct
    }
}
```

See https://github.com/dotnet/csharplang/discussions/9596 and https://github.com/dotnet/runtime/issues/118517

**Expected Behavior**:
Should give warning CS8619; 

**Actual Behavior**:
No warning or error; compiles fine;

**Comment by CyrusNajmabadi** [comment]
ASIDE:

`new ImmutableArray<object?>();` 

This is not actually making a new instance. As this is a struct, you're getting the equivalent of `default`here.

**Comment by CyrusNajmabadi** [comment]
Ok.  Here's my *guess* (but up to @dotnet/roslyn-compiler  to validate).  When you do `new SomeStruct<...>()`, the nullability doesn't actually matter.  That's because you're not actually doing anything (except getting a non-null default instance).   In other words: `new SomeStruct<X>()` is equivalent to `default` (as long as `X` is runtime compatible with the dest type).  As such, there's no actual effective NRT checking that can be done here.  And if we warned here, we'd have to warn for `MyStruct<object> c = default;` which i believe we have chosen not to do.

The reason for this is that `= default` is totally legit for many structs, to put them into a state where they are usable and will not cause problems.  So rather than deal with false positives, we opt for false negatives here. 

Again, that's my *guess*, but it at least fits an intuition i have for this space.

**Comment by CyrusNajmabadi** [comment]
@jaredpar @jcouv if you have more info.

**Comment by huoyaoyuan** [comment]
There isn't warning about struct with constructors either:

```csharp

    struct MyStruct1<T> : IMyInterface<T>
    {
        private T? value;
        public MyStruct(T value) => this.value = value;
    }


    struct MyStruct2<T> : IMyInterface<T>
    {
        public MyStruct() { }
    }

IMyInterface<object> a = new MyStruct1<object?>(null);
IMyInterface<object> b = new MyStruct2<object?>();
```

**Comment by CyrusNajmabadi** [comment]
Ah.  curiouser and curiouser!

**Comment by Brondahl** [comment]
I don't think it has anything to do with constructors or default objects - the same problem happens if you do it with parameters in a method

**Comment by Brondahl** [comment]
Added to the [.NET lab](https://lab.razor.fyi/#jVPdbtMwFBYIJvAV4gms3pCgLmNc8Nct05Zu09C6VWRjEggJNz1NDY4d-dgtGdpDcMsFXHDLG_BQvAJK0qZNV6rlxvbx93N-HPJnjZCuVrFmiRfhw99rFrmMaZihgcQLlBAQGa4keocgQfOoRf6LOEoSa1hPQIuQSDBEOlEmXwmllHJpQA9YBPSokx1ND1vKGnrmt0iBQaNtZGgnC4vN1plPX9XhFbS06GRBvq4ComGGR3SkeJ92GJeOW4TLpPKvnk7vE0TGp4xuUwnjyqCM7_iO26IbG7StAGnMR0DHTMu8I0H44tnmy9WivUp0Ut8SVfnI1IWdnjUUh8qKvktm-scczdakwz6V6sQKgRP9-bup-LWMm7QHEbMIJdynHMumejOTaqZ1t0RpOKk5VsBdrVm26H2ipqatyrTOyM3L4RfmV2WhCwOlwZDJGKYTd6rh7_iUy9Qal25PdrnrRWlZlrNCr3przuzdLVec1eGRK_L4191AyQGPvSAcMp12mUY4TYu_wVHlmtMn2yID74Kb4TGTsWUxvAWNXEln4ex1NYw4jN0Zo6sh1SoCRKXDLOkpgU6jvb93ftiYQx0AM1YDOkXofTVGCWOnYRHWtUKRyXWjPoPkl6AbTdow2kLDbdbBB1zAHkPoTwa5CPzgErdFSK38QCUpFyyv9SZN2BVCjc8lsgE4ufRcHfnDyl9GoKSBL2aqtjzs7cs8OEfPLxJ-WWRyDCMQzrWI14aejec4YQoRH_CozVksFRoeTV2X9TIIN58_2Ww06RtIlTYzkhfaNNWAuNjQgvH0Box6Z_cTXpW_qpk5rgOG9Zlhp1JkzoAJhELq9e0I392_9-P735_fBg_ufLz1Dw):
```
    IMyInterface<T> ChangeClass<T>(MyClass<T?> input) => input; //Warning.
    
    IMyInterface<T> ChangeStruct<T>(MyStruct<T?> input) => input; //No warning.
```

**Comment by Brondahl** [comment]
From the previous conversation in dotnet/csharplang ... do I correctly understand that:

```
System.Console.WriteLine(typeof(List<string?>).IsAssignableTo(typeof(List<string>))); // outputs 'true'.
```

Is known and expected - that the reftype nullability is entirely ignored by actual execution of code?

**Comment by CyrusNajmabadi** [comment]
> Is known and expected - that the reftype nullability is entirely ignored by actual execution of code?

Yes.  Absolutely.  NRT is *intentionally* designed to not have any runtime impact whatsoever.  This was critical for adoption as very few were willing to take on a system that then slowed down existing programs.  It's entirely a static, compile-time linter system, and has been since introduction.
